### ISSUE TITLE
Bulk Edit Products: variant updates ignored when error

### ISSUE THREAD

**Issue by dacook** [openfoodfoundation/openfoodnetwork#11748]
## Description
On the current (old) Bulk Edit Products screen,
When there is an error on a product or its associated variants, changes to the variants are ignored and sometimes lost.

This was found while working on the new "v3" [Bulk Edit Products screen](https://github.com/openfoodfoundation/openfoodnetwork/issues/11059), which is currently blocked by this.

![Screen Shot 2023-11-02 at 1 10 16 pm](https://github.com/openfoodfoundation/openfoodnetwork/assets/4188088/012df555-12ff-4325-8297-68e515d2d7f6)
![Screen Shot 2023-11-02 at 1 10 36 pm](https://github.com/openfoodfoundation/openfoodnetwork/assets/4188088/9a58be8f-acb6-4f55-a7ae-0f8cd953ca0b)

## Expected Behavior
All modified variants should be validated, and saved if valid.

## Actual Behaviour
A. In the case of product error, the variant values remain on-screen, but are not validated. (This is not so bad, because after fixing the product error and trying again, the variants will attempt to save on second try.)
https://github.com/openfoodfoundation/openfoodnetwork/assets/4188088/47694d5f-e20f-4f71-8559-60701797bc58

B. In the case of product success, but variant error, the errored variant values are cleared, with no error message explaining why.
https://github.com/openfoodfoundation/openfoodnetwork/assets/4188088/14e1aa73-ee06-4871-8b49-724c176a4b60

## Steps to Reproduce
Description for scenario B as pictured above:

1. Visit https://staging.openfoodnetwork.org.au/admin/products with feature toggle [admin_style_v3](https://staging.openfoodnetwork.org.au/admin/feature-toggle/features/admin_style_v3) OFF
2. Find or prepare a product with two variants
3. [optional] Edit the product
4. Edit the variant to create an error (eg. name more than 255 chars, or non-numeric unit)
5. Click "Save changes"

* "Changes saved." Variant is reset to past values with no error message.

## Animated Gif/Screenshot
As above

## Workaround
Don't enter invalid variant values

## Your Environment

* Version used: [v4.4.17](https://github.com/openfoodfoundation/openfoodnetwork/releases)
* Browser name and version: Chrome 118.0.5993.117
* Operating System and version (desktop or mobile): macOS

## Possible Fix
1. Update ProductSet to save using nested attributes 
   - requires refactoring of variant stock to allow setting stock on creation (['You need to save the variant to create a stock item before you can set stock levels.'](https://github.com/openfoodfoundation/openfoodnetwork/blob/d8cc45c5c3f84033c99a9a2b631d4c74be346022/app/models/concerns/variant_stock.rb#L140-L143))
2. Update ProductSet to validate all products and their (new) variants first, before saving products and variants. Probably requires some ugly hacks.

**Comment by RachL** [comment]
@dacook as this is on the legacy product page, I'm closing it, please reopen it if I misunderstood it :+1:

**Comment by dacook** [comment]
I'm afraid the new screen uses the same code. I think the issue has improved slightly, but there are still problems, which I don't think are raised elsewhere.
I'll raise a new issue.
